File: src/desktop/apple-script-whitelist.ts
/**
 * AppleScript 白名单（parse5 §3.5.1 + §4.4 注入防御层 1 + INV-27）
 *
 * ## 本文件在 3 层纵深防御中的角色
 *
 *   层 1（TS，本文件 + AppleScriptProvider）：
 *      - typed action enum（AppleScriptActionName）入口
 *      - arg 形状校验：params 的 key 必须是 allowedParams 的子集
 *      - **不内联任何 AppleScript 脚本字符串**（脚本字面量只在 Rust 端）
 *   层 2（Rust, applescript.rs）：二次校验 action 在白名单 + params key 在 allowedParams
 *   层 3（编译期, rust-helper/src/applescript_whitelist.rs）：顶级 const manifest，
 *      编译进 binary；运行时不可改（INV-27 anti-gaming，类比 INV-14）
 *
 * ## 为什么 TS 端不内联脚本
 *
 *   若 TS 端持有脚本字面量，则 LLM 通过 channel 改 env / config / runtime 配置
 *   就能间接改脚本（INV-14 anti-gaming 衍生）。把脚本字面量**只**放在 Rust 端的
 *   顶级 const（编译进 binary），TS 端只镜像 action_name → allowedParams 形状，
 *   是 F3.10.8 typed action enum 红线的可执行形式。
 *
 *   镜像纪律：本文件 `APPLE_SCRIPT_WHITELIST` 与
 *   `rust-helper/src/applescript_whitelist.rs::WHITELIST_STATIC` 字段对齐
 *   （action 名 + allowedParams 名）；新加 action 两边各加一行（≤2 处改动守 02 §4）。
 *
 * ## INV-27 红线（本文件 TS 侧）
 *
 *   - 顶级 const，不从 config / env 读（grep `process.env.APPLE_SCRIPT` 命中即🔴）
 *   - 不导入任何 config / provider-registry / env-reader 模块
 *   - 与 INV-14 同范式：anti-gaming —— LLM 不能通过运行时配置绕过白名单
 *
 * 借鉴：parse5 §3.5.1 + §4.4；F3.10.8 typed action enum；mac-mcp OSAKit 安全路径。
 */
// INV-26 守护：本文件不 import 任何 channel internal / config / subprocess 模块。
// 纯类型 + 顶级 const，零运行时依赖（除 TypeScript 标准类型）。

// ============================================================
// typed action enum
// ============================================================
/**
 * AppleScript typed action 名（与 rust-helper applescript_whitelist.rs 镜像）。
 *
 * 命名规则：lowercase_snake_case（与 Rust 端 naming test 对齐）。
 * 加新 action = 加这里一行 + Rust 端镜像一行（≤2 处改动）。
 */
export type AppleScriptActionName =
  // Finder
  | "finder_new_folder"
  | "finder_empty_trash"
  | "finder_count_windows"
  // Mail
  | "mail_new_message"
  // Safari
  | "safari_open_location"
  | "safari_get_url"
  // Notes
  | "notes_new_note"
  // System / shell（不依赖特定 app 的 AppleEvents）
  | "system_get_volume"
  | "system_get_uptime";

// ============================================================
// 白名单形状
// ============================================================
/**
 * 单条白名单形状（TS 端**只**校验形状，不持脚本）。
 *
 * - `allowedParams`：允许的参数名只读集合；运行时传入的 params 的 key 必须是
 *   它的子集，否则 AppleScriptProvider 直接 outcome=didnt + 拒绝下传 Rust
 *   （层 1 防御；层 2 Rust 端再独立校验，纵深防御）
 *
 * 注：脚本字面量（`script` 字段）**只在** rust-helper/src/applescript_whitelist.rs
 * 出现；本接口刻意不含 `script` 字段（grep 锚点，INV-27 anti-gaming 衍生）。
 */
export interface AppleScriptWhitelistEntry {
  readonly allowedParams: readonly string[];
}

/**
 * 顶级 const 白名单（INV-27 anchor）。
 *
 * 与 `rust-helper/src/applescript_whitelist.rs::WHITELIST_STATIC` 字段镜像：
 * 每条 action 的 allowedParams 与 Rust 端 allowed_params 完全一致。
 *
 * 修改纪律：改这里必须同步改 Rust 端；CI 单测 apple-script-whitelist.spec.ts
 * + Rust applescript_whitelist.rs tests 双向守护形状不漂移。
 */
export const APPLE_SCRIPT_WHITELIST: Record<
  AppleScriptActionName,
  AppleScriptWhitelistEntry
> = {
  // ------------------------------------------------------------------
  // Finder
  // ------------------------------------------------------------------
  finder_new_folder: { allowedParams: [] },
  finder_empty_trash: { allowedParams: [] },
  finder_count_windows: { allowedParams: [] },
  // ------------------------------------------------------------------
  // Mail
  // ------------------------------------------------------------------
  mail_new_message: { allowedParams: ["subject", "content"] },
  // ------------------------------------------------------------------
  // Safari
  // ------------------------------------------------------------------
  safari_open_location: { allowedParams: ["url"] },
  safari_get_url: { allowedParams: [] },
  // ------------------------------------------------------------------
  // Notes
  // ------------------------------------------------------------------
  notes_new_note: { allowedParams: ["name", "body"] },
  // ------------------------------------------------------------------
  // System / shell
  // ------------------------------------------------------------------
  system_get_volume: { allowedParams: [] },
  system_get_uptime: { allowedParams: [] },
};

// ============================================================
// 公共查询 API（AppleScriptProvider + 单测用）
// ============================================================
/**
 * 是否为已知 typed action（层 1 入口校验）。
 *
 * 用 `name in APPLE_SCRIPT_WHITELIST` 而非 Object.hasOwn 是为了兼容性 +
 * 让 TypeScript narrowed 到对应的 whitelist entry。
 */
export function isKnownAction(name: string): name is AppleScriptActionName {
  return Object.prototype.hasOwnProperty.call(APPLE_SCRIPT_WHITELIST, name);
}

/**
 * 校验 params 的所有 key 是否都是 allowedParams 的子集。
 *
 * @returns 第一次违规的 key（caller 写入 error 信息）；全部合规返 null。
 */
export function findDisallowedParamKey(
  action: AppleScriptActionName,
  params: Record<string, unknown>,
): string | null {
  const allowed = APPLE_SCRIPT_WHITELIST[action].allowedParams;
  for (const k of Object.keys(params)) {
    if (!allowed.includes(k)) return k;
  }
  return null;
}

/**
 * 白名单全部 action 名（单测覆盖用 + doctor 自检用）。
 *
 * 返只读数组；不允许调用方修改（INV-27 anti-gaming 衍生）。
 */
export function allWhitelistActions(): readonly AppleScriptActionName[] {
  return Object.keys(APPLE_SCRIPT_WHITELIST) as AppleScriptActionName[];
}

File: rust-helper/src/applescript_whitelist.rs
//! AppleScript whitelist manifest (parse5 §3.5.1 + §4.4 + INV-27).
//!
//! ## 3 层纵深防御（depth-in-defense）
//!
//!   层 1 (TS)：`src/desktop/apple-script-whitelist.ts` 顶级 const + AppleScriptProvider
//!              校验 action 在白名单 + params key 在 allowedParams
//!   层 2 (Rust)：`applescript::run` 二次校验 action + allowedParams（不信任 TS 端）
//!   层 3 (compile-time)：**本文件** 顶级 const，编译进 binary；运行时不可改
//!              （防 LLM 通过 channel 改 env 绕过；类比 INV-14 anti-gaming）
//!
//! ## INV-27 红线
//!
//!   - 本文件是**顶级 const**，从不读 env / config / 文件系统
//!   - 不存在任何 `std::env` 系列调用（var / vars / var_os 均禁；自测源码扫描）
//!   - 加新 action = 加这里一行 + TS 端镜像一行（≤2 处改动守 02 §4）
//!
//! ## 模板渲染
//!
//!   每个模板含占位符 `${param_name}`；运行时由 `applescript::render_template`
//!   做严格 string-replace。**不做 shell escape**：因为层 4（子进程 argv 数组）
//!   保证参数不经 shell interpolation；allowedParams 白名单限制输入维度。
//!   此外，参数值再经 `validate_param_value` 过滤（仅允许字母数字 + 限标点），
//!   即使模板拼接出脚本，osascript 也只看 argv[1]，不存在 command injection。
//!
//! ## 选型：subprocess `/usr/bin/osascript`
//!
//!   见 `examples/smoke_applescript.rs` 决策：
//!   - 零依赖（无 osakit crate / FFI 风险）
//!   - argv 数组传参（非 shell string），天然无注入
//!   - 错误码 + stderr 提供 clean error_kind 映射

use std::collections::HashMap;
use std::sync::OnceLock;

/// AppleScript 模板（对应 parse5 §3.5.1 AppleScriptTemplate）。
///
/// - `script`: 含 `${param}` 占位符的 AppleScript 文本
/// - `allowed_params`: 允许的参数名白名单（防注入层 2；运行时 params 的 key
///   必须是 allowed_params 的子集，否则 applescript::run 拒绝）
#[derive(Debug, Clone)]
pub struct AppleScriptTemplate {
    pub script: &'static str,
    pub allowed_params: &'static [&'static str],
}

/// 静态 manifest：action_name → 模板。
///
/// 覆盖 6 个常见 typed action（parse5 §3.5.1 枚举）。新加 action 加这里一行 +
/// TS 端 apple-script-whitelist.ts 镜像一行。
///
/// 注意：
///   - `do shell script "..."` 内部不做插值（参数不进 shell 路径）
///   - 所有 `${...}` 占位符必须出现在 allowed_params 中（编译期手工保证）
pub const WHITELIST_STATIC: &[(&str, AppleScriptTemplate)] = &[
    // ------------------------------------------------------------------
    // Finder
    // ------------------------------------------------------------------
    (
        "finder_new_folder",
        AppleScriptTemplate {
            script: "tell application \"Finder\" to make new folder at (path to desktop folder)",
            allowed_params: &[],
        },
    ),
    (
        "finder_empty_trash",
        AppleScriptTemplate {
            script: "tell application \"Finder\" to empty trash",
            allowed_params: &[],
        },
    ),
    (
        "finder_count_windows",
        AppleScriptTemplate {
            script: "tell application \"Finder\" to count of windows",
            allowed_params: &[],
        },
    ),
    // ------------------------------------------------------------------
    // Mail
    // ------------------------------------------------------------------
    (
        "mail_new_message",
        AppleScriptTemplate {
            script: "tell application \"Mail\"\nmake new outgoing message with properties {visible:true, subject:${subject}, content:${content}}\nend tell",
            allowed_params: &["subject", "content"],
        },
    ),
    // ------------------------------------------------------------------
    // Safari
    // ------------------------------------------------------------------
    (
        "safari_open_location",
        AppleScriptTemplate {
            script: "tell application \"Safari\" to open location ${url}",
            allowed_params: &["url"],
        },
    ),
    (
        "safari_get_url",
        AppleScriptTemplate {
            script: "tell application \"Safari\" to get URL of front document",
            allowed_params: &[],
        },
    ),
    // ------------------------------------------------------------------
    // Notes
    // ------------------------------------------------------------------
    (
        "notes_new_note",
        AppleScriptTemplate {
            script: "tell application \"Notes\"\nmake new note with properties {name:${name}, body:${body}}\nend tell",
            allowed_params: &["name", "body"],
        },
    ),
    // ------------------------------------------------------------------
    // System Settings / shell（不依赖特定 app 的 AppleEvents）
    // ------------------------------------------------------------------
    (
        "system_get_volume",
        AppleScriptTemplate {
            script: "output volume of (get volume settings)",
            allowed_params: &[],
        },
    ),
    (
        "system_get_uptime",
        AppleScriptTemplate {
            script: "do shell script \"uptime\"",
            allowed_params: &[],
        },
    ),
];

/// 运行时只读 HashMap 视图（OnceLock 构造一次，永久复用）。
///
/// 对外只暴露查询 API；不存在写入 API（INV-27 anti-gaming 守护）。
pub fn whitelist() -> &'static HashMap<&'static str, AppleScriptTemplate> {
    static MAP: OnceLock<HashMap<&'static str, AppleScriptTemplate>> = OnceLock::new();
    MAP.get_or_init(|| {
        let mut m = HashMap::with_capacity(WHITELIST_STATIC.len());
        for (name, tpl) in WHITELIST_STATIC {
            m.insert(*name, tpl.clone());
        }
        m
    })
}

/// 查询 action 是否在白名单 + 返模板引用。
///
/// 注：`whitelist()` 返 `&'static HashMap` 但 borrow 类型是 `&'_`；
/// 因为 underlying storage 是 OnceLock 持有 `&'static str` key +
/// clone 的 `AppleScriptTemplate`，无法直接返 `&'static AppleScriptTemplate`。
/// caller (`applescript::run`) 在函数作用域内用，无需 'static lifetime。
pub fn lookup(action: &str) -> Option<AppleScriptTemplate> {
    whitelist().get(action).cloned()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn whitelist_static_has_expected_actions() {
        let names: Vec<&str> = WHITELIST_STATIC.iter().map(|(n, _)| *n).collect();
        assert!(names.contains(&"finder_new_folder"), "missing finder_new_folder");
        assert!(names.contains(&"mail_new_message"), "missing mail_new_message");
        assert!(names.contains(&"safari_open_location"), "missing safari_open_location");
        assert!(names.contains(&"notes_new_note"), "missing notes_new_note");
        assert!(names.contains(&"system_get_uptime"), "missing system_get_uptime");
    }

    #[test]
    fn whitelist_lookup_returns_template_for_known_action() {
        let tpl = lookup("finder_count_windows").expect("known action");
        assert!(tpl.script.contains("count of windows"));
        assert!(tpl.allowed_params.is_empty());
    }

    #[test]
    fn whitelist_lookup_returns_none_for_unknown_action() {
        assert!(lookup("do_shell_script_rm_rf").is_none());
        assert!(lookup("").is_none());
    }

    #[test]
    fn whitelist_immutable_top_level_const_no_env_var_read() {
        // INV-27 守护：源码扫描确保本文件不出现真实 env::var 调用（带括号 / 路径）。
        // 拼接 pattern 避免本测试自身字符串污染扫描（self-reference）。
        // 锚点：若未来引入真实 env 读，应同时改本测试 + INV-27。
        let src = include_str!("applescript_whitelist.rs");
        let env_var_call: String = ["std", "env", "var"].join("::") + "(";
        let env_vars_call: String = ["std", "env", "vars"].join("::") + "(";
        let env_var_os_call: String = ["std", "env", "var_os"].join("::") + "(";
        assert!(
            !src.contains(&env_var_call),
            "INV-27 violated: real env read call found"
        );
        assert!(
            !src.contains(&env_vars_call),
            "INV-27 violated: real env::vars call found"
        );
        assert!(
            !src.contains(&env_var_os_call),
            "INV-27 violated: real env::var_os call found"
        );
        // 顶级 const manifest 必须存在（INV-27 anchor）
        assert!(
            src.contains("WHITELIST_STATIC"),
            "INV-27 anchor: WHITELIST_STATIC top-level const must exist"
        );
    }

    #[test]
    fn whitelist_all_templates_have_consistent_placeholder_and_allowed_params() {
        // 编译期手工保证的"占位符 ⊆ allowed_params"在测试期再验一次。
        for (name, tpl) in WHITELIST_STATIC {
            let mut placeholders = Vec::new();
            let bytes = tpl.script.as_bytes();
            let mut i = 0;
            while i < bytes.len() {
                if bytes[i] == b'$' && i + 1 < bytes.len() && bytes[i + 1] == b'{' {
                    let start = i + 2;
                    if let Some(end_rel) = tpl.script[start..].find('}') {
                        let ph = &tpl.script[start..start + end_rel];
                        placeholders.push(ph);
                        i = start + end_rel + 1;
                        continue;
                    }
                }
                i += 1;
            }
            for ph in placeholders {
                assert!(
                    tpl.allowed_params.iter().any(|p| *p == ph),
                    "action '{}' has placeholder '{}={{{}}}' not in allowed_params",
                    name,
                    '$',
                    ph
                );
            }
        }
    }

    #[test]
    fn whitelist_action_names_are_lowercase_snake_case() {
        // 防 typo：action 名加错成 "Finder_New_Folder" 之类（TS 端 zod enum 也守）。
        for (name, _) in WHITELIST_STATIC {
            assert!(
                name.chars()
                    .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_'),
                "action name '{}' not lowercase_snake_case",
                name
            );
        }
    }
}
